datastream-server-restli/src/test/java/com/linkedin/datastream/server/TestCoordinator.java

```diff
@@ -656,6 +657,86 @@ public void testCoordinationWithStickyMulticastStrategy() throws Exception {
     zkClient.close();
   }
 
+  /**
+   * testCoordinationWithStickyMulticastStrategyAndMaxTaskLimit is a test to verify that assignment cannot complete
+   * when the tasks per instance is exceeded by the configured threshold
+   * <ul>
+   *     <li>Create a coordinator with CoordinatorConfig.CONFIG_MAX_DATASTREAM_TASKS_PER_INSTANCE set to 5</li>
+   *     <li>create 3 instances and 4 datastreams, verify assignment cannot complete</li>
+   *     <li>create another instance, verify assignment completes now</li>
+   * </ul>
+   */
+  @Test
+  public void testCoordinationWithStickyMulticastStrategyAndMaxTaskLimit() throws Exception {
+    String testCluster = "testCoordinationMaxTasksPerInstance";
+    String testConnectorType = "testConnectorType";
+    Properties properties = new Properties();
+    // Set max tasks per instance to 5, since some instances will get 6 tasks if 4 datastreams with 4 tasks each are
+    // created across 3 instances. With 4 instances, the tasks per instance will be less than 5.
+    properties.put(CoordinatorConfig.CONFIG_MAX_DATASTREAM_TASKS_PER_INSTANCE, "5");
+    Coordinator instance1 = createCoordinator(_zkConnectionString, testCluster, properties);
+
+    TestHookConnector connector1 = new TestHookConnector("connector1", testConnectorType);
+    //Question why the multicast strategy is within one coordinator rather than shared between list of coordinators
+    instance1.addConnector(testConnectorType, connector1, new StickyMulticastStrategy(Optional.of(4), Optional.of(2)), false,
+        new SourceBasedDeduper(), null);
+    instance1.start();
+
+    Coordinator instance2 = createCoordinator(_zkConnectionString, testCluster);
+    TestHookConnector connector2 = new TestHookConnector("connector2", testConnectorType);
+    instance2.addConnector(testConnectorType, connector2, new StickyMulticastStrategy(Optional.of(4), Optional.of(2)), false,
+        new SourceBasedDeduper(), null);
+    instance2.start();
+
+    Coordinator instance3 = createCoordinator(_zkConnectionString, testCluster);
+    TestHookConnector connector3 = new TestHookConnector("connector3", testConnectorType);
+    instance3.addConnector(testConnectorType, connector3, new StickyMulticastStrategy(Optional.of(4), Optional.of(2)), false,
+        new SourceBasedDeduper(), null);
+    instance3.start();
+
+    ZkClient zkClient = new ZkClient(_zkConnectionString);
+    List<TestHookConnector> connectors = new ArrayList<>();
+    connectors.add(connector1);
+    connectors.add(connector2);
+    connectors.add(connector3);
+    List<String> datastreamNames = ImmutableList.of("datastream1", "datastream2", "datastream3", "datastream4");
+
+    for (String name : datastreamNames) {
+      DatastreamTestUtils.createAndStoreDatastreams(zkClient, testCluster, testConnectorType, name);
+    }
+    boolean assignmentComplete = waitTillAssignmentIsComplete(16, WAIT_TIMEOUT_MS,
```

Comment:
Is there are way to make this test more unit-test like? Just test `validateNewAssignment` I mean? My concern is that if assignment fails for any other reason, it would still look the same for the reader of waitTillAssignmentIsComplete's output even though it actually did not fail due to max tax limit if I am reading this correct

Reply:
Added a new unit test which is more like a unit test. Wanted to keep this one, since we are testing that increasing the instances allows the assignment to complete.